### PR TITLE
feat(cli): add --puppeteer-script flag to collect

### LINKED ISSUE #89: Support for `--puppeteer-script`
Usage:

`lhci collect --puppeteer-script=./path/to/script.js`

```js
// This function is invoked before the first run of each URL under test
module.exports = async (browser, urlUnderTest, allCollectOptions) => {
  if (!urlUnderTest.includes('/app')) return;

  const page = await browser.newPage();
  await page.goto('http://localhost:3000/login');
  await page.type('#username', 'test_user');
  await page.type('#password', 'password');
  await page.click('button');
  await page.waitFor('.welcome-logged-in-user');
}
```

### PR DESCRIPTION
**Pull request by patrickhulce**
closes #89 

Usage looks like:

```
lhci collect \
  --start-server-command="yarn serve" \
  --url=http://localhost:8080/ \
  --puppeteer-script=./path/to/login-with-puppeteer.js
```

**Example `login-with-puppeteer.js`**

```js
/**
 * @param {puppeteer.Browser} browser
 * @param {{url: string, options: LHCI.CollectOptions}} context
 */
module.exports = async (browser, context) => {
  const page = await browser.newPage();
  await page.goto('http://localhost:8080/login');
  await page.type('#username', 'admin');
  await page.type('#password', 'password');
  await page.click('[type="submit"]');
  await page.waitForNavigation();
};
```

### CODE REVIEW

File: docs/cli.md [hunk 1]
```diff
@@ -127,8 +131,36 @@ lhci collect --numberOfRuns=5 --url=https://example.com
...
+  const page = await browser.newPage();
+  await page.goto('http://localhost:8080/login');
+  await page.type('#username', 'admin');
+  await page.type('#password', 'password');
+  await page.click('[type="submit"]');
+  await page.waitForNavigation();
+};
+```
+
+Lighthouse CI will then use this browser that the script sets up when running Lighthouse. Note that if you store your credentials in `localStorage` or anything other than a cookie you might want to pair this option with `--settings.disableStorageReset` to force Lighthouse to keep the cache state.
```
Comment (connorjclark) [inline docs/cli.md]: Link to our auth docs too?

File: docs/recipes/docker-server/update-dockerhub.sh [hunk 2]
```diff
@@ -1,8 +1,7 @@
 #!/bin/bash
 
 CURRENT_VERSION=$(node -e "console.log(require('./package.json').dependencies['@lhci/server'])")
-# NEXT_VERSION=$(yarn info @lhci/server | grep 'latest:' -A 1 | tail -n 1 | grep -o "'.*'" | sed s/\'//g)
-NEXT_VERSION=0.3.3
+NEXT_VERSION=$(yarn info @lhci/server | grep 'latest:' -A 1 | tail -n 1 | grep -o "'.*'" | sed s/\'//g)
```
Comment (connorjclark) [inline docs/recipes/docker-server/update-dockerhub.sh]: woah. maybe do a call to node like the line above? this is inscrutable :)

File: docs/cli.md (hunk 1 above)
Comment (connorjclark) [reply docs/cli.md]: oh, it's linked in troubleshooting.

File: packages/cli/src/collect/puppeteer-manager.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  /** @return {boolean} */
+  isActive() {
+    return !!this._options.puppeteerScript;
+  }
+
+  /** @return {Promise<number>} */
+  async getBrowserPort() {
+    const browser = await this._getBrowser();
+    const [_ = '', port = ''] = browser.wsEndpoint().match(/:(\d+)/) || [];
```
Comment (connorjclark) [inline packages/cli/src/collect/puppeteer-manager.js]: can you do new URL().port instead here?

File: docs/cli.md [hunk 4]
```diff
@@ -127,8 +131,36 @@ lhci collect --numberOfRuns=5 --url=https://example.com
...
+When running Lighthouse CI on a page with behind authentication, you'll need to authorize the browser that Lighthouse CI will be using. While there are [several](./configuration.md#page-behind-authentication) different [options](https://github.com/GoogleChrome/lighthouse/blob/v5.6.0/docs/authenticated-pages.md) to accomplish this, `--puppeteer-script` is one of the most flexible and convenient.
+
+`--puppeteer-script` accepts a path to a JavaScript file that exports a function that will be invoked by Lighthouse CI before navigating, e.g. a script that will login to your site.
+
+**Example `puppeteer-script.js`**
+
+```js
+/**
+ * @param {puppeteer.Browser} browser
+ * @param {{url: string, options: LHCI.CollectOptions}} context
```
Comment (connorjclark) [inline docs/cli.md]: I don't think this type is defined anywhere. is it `{url:string, options: LHCI.Options}`?

File: docs/recipes/docker-server/update-dockerhub.sh (hunk 2 above)
Comment (patrickhulce) [reply docs/recipes/docker-server/update-dockerhub.sh]: doing it all in node would be at least a dozen lines though :)

File: packages/cli/src/collect/puppeteer-manager.js (hunk 3 above)
Comment (patrickhulce) [reply packages/cli/src/collect/puppeteer-manager.js]: good call

File: docs/cli.md (hunk 4 above)
Comment (patrickhulce) [reply docs/cli.md]: yeah it's not it was meant to just be user friendly "here are the collect options" but i'll use the real one :)

**Pull request merged**